src/main/java/org/example/EchoMessageBus.java

```diff
@@ -0,0 +1,44 @@
+package org.example;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.net.Socket;
+
+public class EchoMessageBus extends Thread {
+    private final Socket clientSocket;
+    private final InputOutput io;
+
+    public EchoMessageBus(Socket socket) {
+        this.clientSocket = socket;
+        io = new InputOutput();
+    }
+
+    public void run() {
+        try {
+            System.out.println("[CLIENT CONNECTED]");
+            BufferedReader in = io.getSocketInputStream(clientSocket);
+            PrintWriter out = io.sendSocketOutputStream(clientSocket);
+            String userInput;
+            while((userInput = io.readClientInputStream(in)) != null) {
+                System.out.printf("[FROM CLIENT] %s%n", userInput);
+                if (userInput.equals("end")) {
+                    publishMessage(out, "[SHUTTING DOWN ECHO SERVER]");
+                    System.out.println("[SHUTTING DOWN ECHO SERVER]");
+                    this.clientSocket.close();
+                    return;
+                }
+                publishMessage(out, userInput);
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```

Comment:
One of the error handling best-practices when we develop webservers is Let it fail fast. Meaning that we don't catch any errors inside classes. If we want to wrap errors to not expose error details outside, we'd have a single try-catch on a highest possible level which catches all the errors from workers that execute and does all the required wrapping. Propose to follow that principle here as well. And even if we don't need special error handling for this project - good! For maintenence less code is always better then more code

Reply:
This comment was very helpful for untangling the refactoring issue I was having. There is a new commit with moves the try/catch to the highest possible level.

Reply:
Sounds awesome!